### ISSUE TITLE
[Task 28] Decouple Algorithm Configuration from the Training Pipeline and Add GDPO

### ISSUE THREAD

**Issue by curious-whq** [redai-infra/Relax#105]
## Task

I would like to work on Advanced Task 28: decouple algorithm configuration from the training pipeline and integrate GDPO through a unified algorithm registry.

## Repository and Development Baseline

| Item | Value |
| --- | --- |
| Upstream repository | https://github.com/redai-infra/Relax |
| Development repository | https://github.com/curious-whq/Relax |
| Baseline commit | `039ce876d25540adad847d4223b4de4722d8f425` |
| Development branch | `feature/gdpo` |
| Current implementation commit | `c7984d0078f1336581f469f863c2b07f15a8647f` |

## Environment

| Item | Configuration |
| --- | --- |
| Image | `ghcr.io/redai-infra/relaxrl:dev-20260715-8325919e` |
| Image digest | `sha256:3fa8ce578acda6c829b83016bde42c38fa892681e4f36ca330f545616fe578e2` |
| Python | 3.12.3 |
| CUDA | 12.9 |
| PyTorch | 2.11.0+cu129 |
| Ray | 2.56.0 |
| SGLang | 0.5.12.post1 |
| GPU | 1 × NVIDIA RTX 6000 Ada Generation 48 GB |
| CPU | 1 × 13th Gen Intel(R) Core(TM) i9-13980HX, 24 cores / 32 threads |
| Memory | 1.0 TiB |
| OS | Ubuntu 22.04.5 LTS |

## Background

Algorithm integration currently relies on algorithm-name strings and distributed `if/elif` branches. Adding a new algorithm requires updating multiple locations, which makes it easy to miss configuration validation, reward processing, advantage computation, or loss dispatch.

Relax also does not currently provide the multi-reward normalization required by GDPO:

1. Normalize each reward dimension independently within every prompt group.
2. Combine the normalized advantages from all reward dimensions.
3. Normalize the combined result across the sequence batch.

## Goals

- Introduce a unified algorithm registration and dispatch mechanism.
- Manage algorithm names, capabilities, and implementation hooks through the registry.
- Migrate existing algorithms to the registry without changing their numerical behavior.
- Integrate GDPO through the registry.
- Support at least two configurable reward keys.
- Avoid algorithm-specific branches such as `if algorithm == "gdpo"`.

### 1. Algorithm Registry

Each algorithm will register the following information:

- Algorithm name
- Capability declarations
- Reward extractor
- Reward processor
- Advantage estimator
- Ratio builder
- Policy objective
- Configuration validator
- Service topology

Capabilities will be used by the training orchestration layer to determine:

- Whether a critic is required
- Whether complete reward groups are required
- The minimum number of reward keys
- Whether batch-wise advantage normalization is required
- Whether full log probabilities are required
- Which execution modes are supported

CLI choices, Controller service mappings, reward processing, advantage computation, and loss dispatch will be derived from the registry.

### 2. Existing Algorithm Migration

The following algorithms will be migrated to the registry-driven dispatch mechanism:

- GRPO
- GSPO
- SAPO
- CISPO
- REINFORCE++
- REINFORCE++ baseline
- PPO

The migration will only change how implementations are discovered and dispatched. Existing formulas, default arguments, and numerical behavior will remain unchanged.

### 3. GDPO Reward Processing

Multiple reward dimensions are configured using:

```bash
--advantage-estimator gdpo
--reward-key correctness
--reward-keys correctness format
--reward-weights 1.0 1.0
```

For multiple responses sampled from the same prompt:

1. Normalize each configured reward dimension independently within the prompt group.
2. Combine the normalized values using `--reward-weights`.
3. Normalize the combined sequence-level advantages across the complete batch.
4. Broadcast each sequence-level advantage to the corresponding response tokens.

Missing rewards, non-numeric rewards, NaN/Inf values, incomplete groups, and invalid configurations must fail fast. A zero-variance reward dimension contributes zero to the combined advantage.

### 4. Batch-Normalization Boundary

GDPO requires:

```text
rollout_batch_size * n_samples_per_prompt == global_batch_size
```

This ensures that one rollout produces exactly one sequence-wise normalization batch and that reward groups are not split across normalization batches.

The component path requests a complete `global_batch_size` from TransferQueue. The Megatron path computes global statistics through the data-parallel process group.

### Core Implementation

- `relax/core/registry.py`
- `relax/components/advantages.py`
- `relax/backends/megatron/loss.py`
- `relax/utils/utils.py`
- `relax/utils/arguments.py`
- `relax/utils/training/algorithm_ops.py`
- `relax/engine/filters/dynamic_sampling_filters.py`

### Unit Tests

- `tests/core/test_algorithm_registry.py`
- `tests/utils/test_algorithm_dispatch.py`
- `tests/utils/training/test_algorithm_ops.py`
- `tests/utils/training/test_gdpo.py`
- `tests/examples/test_gdpo_reward.py`

### Examples and Documentation

- `examples/algorithms/gdpo_reward.py`
- `examples/algorithms/run-qwen3-0.6B-1xgpu-gdpo-smoke.sh`
- `examples/algorithms/run-qwen3-4B-8xgpu-gdpo.sh`
- `examples/algorithms/README.md`
- `docs/en/examples/algorithms.md`
- `docs/zh/examples/algorithms.md`
- `docs/en/guide/configuration.md`
- `docs/zh/guide/configuration.md`

## Out of Scope

- Changing the mathematical formulas or default hyperparameters of existing algorithms
- Renaming existing algorithm CLI options
- Adding third-party dependencies
- Changing the overall Ray, Megatron, or SGLang architecture
- Reproducing the full experimental results from the GDPO paper
- Reducing the effective batch size, sequence length, or sample count to make training pass
- Optimizing distributed collective performance

## Minimal Training Example

The minimal example uses:

- **Model:** Qwen3-0.6B
- **Dataset:** GSM8K
- **Execution mode:** 1-GPU colocate
- **Reward dimensions:** `correctness` and `format`

Reward definitions:

- `correctness`: whether the final answer matches the reference answer
- `format`: whether the response follows the required `\boxed{...}` format

Run the smoke test with:

```bash
MODEL_DIR=/path/to/models \
DATA_DIR=/path/to/data \
NUM_ROLLOUT=2 \
bash examples/algorithms/run-qwen3-0.6B-1xgpu-gdpo-smoke.sh
```

Key arguments:

```bash
--advantage-estimator gdpo
--custom-rm-path examples.algorithms.gdpo_reward.reward_func
--reward-key correctness
--reward-keys correctness format
--reward-weights 1.0 1.0
--rollout-batch-size 4
--n-samples-per-prompt 4
--global-batch-size 16
```

### CPU Unit Tests

```bash
python -m pytest -q \
  tests/core/test_algorithm_registry.py \
  tests/utils/test_algorithm_dispatch.py \
  tests/utils/training/test_algorithm_ops.py \
  tests/utils/training/test_gdpo.py \
  tests/examples/test_gdpo_reward.py
```

The tests cover:

- Algorithm registration, duplicate registration, and unknown algorithms
- Registry-driven reward, advantage, and loss dispatch
- Numerical equivalence of existing algorithms
- The GDPO multi-reward formula
- Reward-collapse behavior
- Missing reward keys
- Non-numeric and non-finite rewards
- Zero-variance reward dimensions
- Incomplete reward groups
- Local and distributed batch normalization
- The correctness/format example reward

### Formatting and Static Checks

```bash
pre-commit run --all-files
```

### GPU Smoke Test

```bash
MODEL_DIR=/path/to/models \
DATA_DIR=/path/to/data \
NUM_ROLLOUT=2 \
bash examples/algorithms/run-qwen3-0.6B-1xgpu-gdpo-smoke.sh
```

The smoke test should verify that:

- The Ray job starts and exits successfully.
- At least two training iterations complete.
- `--advantage-estimator gdpo` is parsed and dispatched successfully.
- Both `correctness` and `format` rewards are produced.
- Policy loss and gradient norm remain finite.
- No unexplained errors, NaN/Inf values, OOM errors, or dropped samples appear in the logs.

## Acceptance Criteria

- [ ] Algorithm names, capabilities, and implementations are managed by the registry.
- [ ] Algorithm-name lists are not maintained in multiple locations.
- [ ] Complete algorithm `if/elif` dispatch chains are removed.
- [ ] `--advantage-estimator gdpo` works correctly.
- [ ] GDPO supports at least two configurable reward keys.
- [ ] GDPO is integrated through the registry without `if algorithm == "gdpo"`.
- [ ] Existing algorithm behavior remains numerically equivalent.
- [ ] GDPO formulas and error scenarios are covered by unit tests.
- [ ] A minimal `correctness`/`format` training example is provided.
- [ ] CPU tests pass.
- [ ] `pre-commit run --all-files` passes.
- [ ] The 1-GPU smoke test passes.
- [ ] Algorithm integration and GDPO configuration are documented.
